src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs

```diff
@@ -389,124 +390,109 @@ public KestrelHttpParser(IKestrelTrace log)
                 {
                     sawWhitespace = true;
                 }
+            }
 
-                index++;
-                pCurrent++;
+            if (index == length)
+            {
+                RejectRequest(RequestRejectionReason.NoColonCharacterFoundInHeaderLine);
+            }
+            if (sawWhitespace)
+            {
+                RejectRequest(RequestRejectionReason.WhitespaceIsNotAllowedInHeaderName);
             }
-            return result;
+            return index;
         }
 
-        private unsafe void TakeSingleHeader<T>(byte* pHeader, int headerLineLength, T handler) where T : IHttpHeadersHandler
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe void TakeSingleHeader<T>(byte* headerLine, int length, T handler) where T : IHttpHeadersHandler
         {
-            var nameEnd = -1;
-            var valueStart = -1;
-            var valueEnd = -1;
-            var index = 0;
-            var pCurrent = pHeader + index;
-            var pEnd = pHeader + headerLineLength;
+            // Skip CR, LF from end position
+            var valueEnd = length - 3;
+            var nameEnd = FindEndOfName(headerLine, length);
 
-            nameEnd = IndexOfNameEnd(pHeader, index, headerLineLength);
-
-            if (nameEnd == -1)
+            if (headerLine[valueEnd + 2] != ByteLF)
             {
-                RejectRequest(RequestRejectionReason.NoColonCharacterFoundInHeaderLine);
+                RejectRequest(RequestRejectionReason.HeaderValueMustNotContainCR);
+            }
+            if (headerLine[valueEnd + 1] != ByteCR)
+            {
+                RejectRequest(RequestRejectionReason.MissingCRInHeaderLine);
             }
 
-            // Skip colon
-            index += nameEnd + 1;
-            pCurrent += index;
-            valueStart = index;
-            var pValueStart = pCurrent;
-
-            while (pCurrent < pEnd)
+            // Skip colon from value start
+            var valueStart = nameEnd + 1;
+            // Ignore start whitespace
+            for(; valueStart < valueEnd; valueStart++)
             {
-                var ch = *pCurrent;
+                var ch = headerLine[valueStart];
                 if (ch != ByteTab && ch != ByteSpace && ch != ByteCR)
                 {
-                    valueStart = index;
-                    pValueStart = pCurrent;
                     break;
                 }
                 else if (ch == ByteCR)
                 {
-                    break;
+                    RejectRequest(RequestRejectionReason.HeaderValueMustNotContainCR);
                 }
-                pCurrent++;
-                index++;
             }
 
-            var endIndex = headerLineLength - 1;
-            pEnd--;
 
-            if (*pEnd != ByteLF)
+            // Check for CR in value
+            var i = valueStart + 1;
+            if (Vector.IsHardwareAccelerated)
```

Comment:
Span would as would have to refix the DangerousReference; but it looks like it can be extracted with only a small hit. Improving the tests to make sure everything is covered.